Add service certificate annotations Kuttl tests

kabicin opened the pull request:
- Adds service certificate annotations Kuttl tests to check that annotations can be properly applied to CertManager Certificates with WLO's .spec.service.certificate.annotations

Related issue: https://github.com/WASdev/websphere-liberty-operator/issues/440

idlewis commented:
@kabicin 
1. The test failed in the first step when I ran it. The error was this:
```
        -          secretName: service-with-cert-annotations-liberty-svc-tls-cm
        +          secretName: service-with-cert-annotations-liberty-svc-tls-ocp
```
However, as this test is testing annotations (there is already a more general service certificate test), I don't see the value in checking either the env or the volumes for the deployment. It is liable to make the test more fragile. I think you could probably take this check out rather than fixing it.

1. The original issue said that annotations from the parent CR should propogate down to the certificate (in addition to annotations specifically added to the service). Could you add a test for that?

kabicin commented:
I wasn't sure which test script(s) to validate exactly, but used `scripts/e2e-kind.sh`, so I'm not sure if there is another one you are running that doesn't use certmanager but runs on OCP.

> However, as this test is testing annotations (there is already a more general service certificate test), I don't see the value in checking either the env or the volumes for the deployment. It is liable to make the test more fragile. I think you could probably take this check out rather than fixing it.

Sounds good, I removed check for the deployment. 

> The original issue said that annotations from the parent CR should propogate down to the certificate (in addition to annotations specifically added to the service). Could you add a test for that?

Thanks for catching that, I added that to the tests now.

idlewis commented:
@kabicin 
Test is now passing for me and the other changes look good.
One more thing though :-)
At the end of the test, I see the following:
```
$ oc get secret
NAME                                               TYPE                                  DATA   AGE
builder-dockercfg-v2llw                            kubernetes.io/dockercfg               1      2d19h
builder-token-fghxp                                kubernetes.io/service-account-token   4      2d19h
default-dockercfg-9z7rm                            kubernetes.io/dockercfg               1      2d19h
default-token-cz2x4                                kubernetes.io/service-account-token   4      2d19h
deployer-dockercfg-vjdn9                           kubernetes.io/dockercfg               1      2d19h
deployer-token-hwczf                               kubernetes.io/service-account-token   4      2d19h
service-with-cert-annotations-liberty-svc-tls-cm   kubernetes.io/tls                     3      23s
wlo-ca-tls                                         kubernetes.io/tls                     3      30s
```
I think we should be cleaning up the secret 'service-with-cert-annotations-liberty-svc-tls-cm'. That might need a 04-delete... step, to make sure it is done in the correct order.
I think our tests are not currently very good about cleaning up resources, but we probably should try to make sure that any new tests are better about it.

kabicin commented:
> @kabicin Test is now passing for me and the other changes look good. One more thing though :-) At the end of the test, I see the following:
> 
> ```
> $ oc get secret
> NAME                                               TYPE                                  DATA   AGE
> builder-dockercfg-v2llw                            kubernetes.io/dockercfg               1      2d19h
> builder-token-fghxp                                kubernetes.io/service-account-token   4      2d19h
> default-dockercfg-9z7rm                            kubernetes.io/dockercfg               1      2d19h
> default-token-cz2x4                                kubernetes.io/service-account-token   4      2d19h
> deployer-dockercfg-vjdn9                           kubernetes.io/dockercfg               1      2d19h
> deployer-token-hwczf                               kubernetes.io/service-account-token   4      2d19h
> service-with-cert-annotations-liberty-svc-tls-cm   kubernetes.io/tls                     3      23s
> wlo-ca-tls                                         kubernetes.io/tls                     3      30s
> ```
> 
> I think we should be cleaning up the secret 'service-with-cert-annotations-liberty-svc-tls-cm'. That might need a 04-delete... step, to make sure it is done in the correct order. I think our tests are not currently very good about cleaning up resources, but we probably should try to make sure that any new tests are better about it.

Thanks, I just updated the repos to also delete the cert then the secret, some kind test specific RBAC permission changes were needed to support the additional capability in some operators.
